presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -122,19 +123,19 @@ public Expression rewrite(FieldOrExpression fieldOrExpression)
         }
     }
 
-    public void put(Expression expression, Symbol symbol)
+    public void put(Expression expression, Symbol symbol, Collection<Expression> columnReferences)
     {
         Expression translated = translateNamesToSymbols(expression);
         expressionMappings.put(translated, symbol);
 
-        // also update the field mappings if this expression is a simple field reference
-        if (expression instanceof QualifiedNameReference) {
+        // also update the field mappings if this expression is a field reference
+        if (expression instanceof QualifiedNameReference || (expression instanceof DereferenceExpression && columnReferences.contains(expression))) {
```

Comment:
QNRs should in `columnReferences` too, right? if so, then you can simplify this check to `if (columnReferences.contains(expression))`